### ISSUE TITLE
experiment: test structural representations for v5 semantic discrimination

### ISSUE THREAD

**Issue by lipluscodex** [Liplus-Project/neuron-graph-rag#238]
## 目的

#236 / PR #237 の有効な full-corpus oracle v2 では、既知 v5 development の正解が bge-reranker-base 64位、bge-reranker-v2-m3 45位だった。構造表現とmaximum chunk集約の長文有利を、結果を見る前に固定した2x2 factorial ablationで切り分ける。

## 前提

- v2結果は既知 development 1件の探索的証拠であり、holdoutやproduction性能へ一般化しない。
- v2 evidenceとexact 93文書corpusだけから、chunk数とbest score / rankの相関、正解とtop distractorのwinning chunk、path / title / heading情報をresult-free診断として先に固定する。
- v1/v2のmodel revisions、480 codepoint / 80 overlap、tokenizer max length 512、CPU 4 threadsを継承する。
- body / structural表現は同一のbody chunk境界、offset、本文、hashを使う。structural armのためにwindowやoverlapを変えない。

## 固定arm

4 armを2 representation x 2 aggregationのfactorialとして同一one-shot内で比較する。

1. `body_max`: `chunk_text` + `max(chunk_raw_logits)`。v2基準arm。
2. `structural_max`: 固定structural prefix + `chunk_text` + maximum。
3. `body_nlme`: `chunk_text` + normalized log-mean-exp。
4. `structural_nlme`: 固定structural prefix + `chunk_text` + normalized log-mean-exp。

normalized log-mean-expはtemperature 1.0に固定し、`logsumexp(chunk_raw_logits) - log(chunk_count)` とする。aggregation間で同じrepresentation scoring passのchunk raw logitsを共有する。文書同score時はsource ID昇順で決める。

## structural representation literal

各body chunkへ次のprefixをUTF-8でそのまま付ける。field順、ASCII colon、single ASCII space、LF、最後のblank lineを固定する。

```text
path: {repository_path}
filename: {filename}
title: {document_title}
headings: {heading_chain}

```

- `repository_path`: corpus pathをPOSIX `/` に正規化し、先頭のexact `corpora/github-retrieval-parity-v4/` を1回だけ除く。
- `filename`: POSIX pathのbasename。extensionを残す。
- `document_title`: fenced code外かつcolumn 0で最初に現れる `^# ` の行について、marker後の文字列を正規化する。存在しなければ `[missing]`。
- `heading_chain`: fenced code外かつcolumn 0で `^(#{1,6}) ` に一致し、そのline開始codepointがchunk startより小さいheadingだけを対象にする。level Lを読むたびactive level L以上を除いて追加し、active H1-H6を ` > ` で結ぶ。空なら `[missing]`。
- fenced codeは0〜3 ASCII spacesの後に3個以上のbacktickまたはtildeで開始し、同じ文字かつ開始以上の長さのmarkerで終了する。その内部のheading-like lineは無視する。
- metadata値のwhitespace正規化は、先頭末尾のUnicode whitespaceを除き、内部の連続Unicode whitespaceをsingle ASCII spaceへ置換する。空になった値は `[missing]`。
- prefix capは設けない。body chunk本文は正規化せず、prefixの直後へbyte-identicalに連結する。
- この変換をquery / gold identityに依存せず全93文書の全chunkへ適用する。

## tokenizer truncation diagnostics

prefix追加によるtruncation増加を結果を見る前に監査できるよう、各model / representation / chunkについて次を保存する。

- body chunk codepoint length、prefix codepoint length（bodyは0）。
- prefixだけを `add_special_tokens=false` でtokenizeしたtoken length（bodyは0）。
- query / passage pairを `add_special_tokens=true`、`truncation=false` でtokenizeした512 truncation前token length。
- truncation前token lengthが512を超えるか。

model / representationごとにpair総数、512超過数、prefix codepoint/token lengthのmin / max / meanを保存する。同一body chunkに対しbodyが512以下でstructuralが512超となった件数と、structural armがtruncationを増やしたかのbooleanも保存する。観測後にchunk/window/prefixを変更しない。

## 成功条件

- cutoffは20に固定する。
- primary success: 少なくとも1つのnon-baseline armで、両pinned modelの正解rankが20以内。
- directional evidence: 同じnon-baseline armで、両modelの正解rankが `body_max` よりstrictに改善。
- length-bias attenuation: 同じrepresentation内でNLME armの `abs(Spearman(chunk_count, document_score))` がmaximum armより両modelで低い。
- primary、directional、attenuationは独立に記録し、観測後に定義を変えない。

## 実装と検証

- holdout-absent allowlistだけをfresh WSLC volumeへcopyし、v2 runtime volumeを再利用しない。
- preflightはforbidden path不在、holdout-bearing input 0件、93文書、全arm literal、body chunk同一性、両model revision / file hash、network disabledを確認する。synthetic forwardだけを行い、登録queryは0回とする。
- workerはqueryを読むがgoldを物理的に持たない。finalizerだけがdevelopment-only goldをread-onlyで読む。
- 2 representation scoring pass / modelをsingle one-shot、retry 0で実行し、4 aggregation結果を同じchunk logitsから算出する。
- 全93文書の全chunk raw logits、構造field、正解rank、top distractor、chunk-count相関、runtime / peak RSS、dependency / input / model / output hash、truncation diagnosticsをappend-only evidenceへ保存する。
- CIはmodel downloadまたは登録query推論を行わず、audit / probe / unit testだけを実行する。

## 制約

- v1/v2、v5以前の凍結artifactとevidenceをbyte-for-byte変更しない。
- mixed-stage v5 query / goldまたはholdout-bearing inputを開かない。GitHub RAGへqueryを送らず、共有DBへ接続しない。
- query別・正解文書別の語句追加、boost、weight、projection、aggregationを禁止する。
- 失敗one-shotをretryで置き換えない。
- default検索やproduction rolloutを変更しない。改善候補の統合は別Issueとする。

## 対象ファイル

- 新規result-free diagnostic / frozen manifest / schema
- 新規ablation runner / holdout-absent WSLC wrapper
- 新規tests / append-only evidence / 実験文書
- 必要なtest suite / documentation index登録

## 参考

- #234
- #236 / PR #237
- #230 / PR #233
